facets/core/prop.ts

```diff
@@ -0,0 +1,81 @@
+import { FacetPropType, NullableTypeOfPropType, TypeOfFacetPropType } from './propType';
+
+type IsNeverNull<T> = Extract<T, null> extends never ? true : false;
+
+export type AvailableCascadeStrategyForPropType<PropType extends FacetPropType> =
+	// Any prop type can use the `overwrite` strategy
+	| 'overwrite'
+	// Only props that are objects can object-merge during cascade" {...a, ...b}
+	| (TypeOfFacetPropType<PropType> extends Record<string, any> ? 'merge' : never)
+	// Only props that are arrays can array-merge during cascasde: [...a, ...b]
+	| (TypeOfFacetPropType<PropType> extends any[] ? 'extend' : never);
+
+export type CascadeStrategy = 'overwrite' | 'merge' | 'extend';
+
+export type FacetPropOptions<
+	// The PropType of the prop.
+	PropType extends FacetPropType,
+	// The `rootValue` provided, which may be `null`.
+	// We provide this as a type parameter so that TypeScript will infer its narrowest -- most
+	// specific -- type and preserve that information in the inferred return type. This lets us
+	// determine statically whether a facet prop will ever be null and adjust its TypeScript
+	// definitions accordingly.
+	RootValue extends NullableTypeOfPropType<PropType>,
+	// When we provide a defaultValue to a facet, it may be `null`.
+	DefaultValue = NullableTypeOfPropType<PropType>,
+> = {
+	// The value of this prop at the "root" of PubPub -- in the absence of any other facet instances
+	// that override it, what value does this prop have?
+	rootValue: RootValue;
+	// The default value of this prop, if not explicitly provided to createFacet(). By contrast with
+	// `rootValue`, this value will actually be stored on a facet instance.
+	defaultValue?: DefaultValue;
+	// Explains how this prop should cascade from higher to lower scopes.
+	cascade?: AvailableCascadeStrategyForPropType<PropType>;
+	// A human-readable label for this prop
+	label?: string;
+};
+
+export type FacetProp<
+	PropType extends FacetPropType = FacetPropType,
+	RootValue extends NullableTypeOfPropType<PropType> = NullableTypeOfPropType<PropType>,
```

Comment:
I love how easy this reads. But I do sometimes prefer generic type parameters shortened to a single letter or otherwise distinguished from module-level types.

Reply:
That's interesting and I agree that the length of the type's name does tend to provide this signal. There are a _lot_ of functions parameterized over these facet types coming up in future PRs, so we should probably be consistent about it.

Reply:
I think we should keep them long for now, but maybe we could prefix them (yuuuuck) with a `$` or `_` or something? Really ugly, but solves both problems...